File: trunk/Yuki/Examples/Communication/CSharp/Src/Server/Servers/BinarySocketServer.cs
using System;
using System.Collections.Generic;
using System.Collections.Concurrent;
using System.Linq;
using System.Net;
using Communication;
using Communication.BaseSystem;
using Communication.Net;
using Communication.Binary;
using Communication.Json;
using Server.Services;

namespace Server
{
    /// <summary>
    /// 本类的所有非继承的公共成员均是线程安全的。
    /// </summary>
    public class BinarySocketServer : TcpServer<BinarySocketServer, BinarySocketSession>
    {
        private ServerImplementation si;
        private JsonLogAspectWrapper<SessionContext> law;
        public BinaryServer<SessionContext> InnerServer { get; private set; }
        public ServerContext ServerContext { get; private set; }

        private int MaxBadCommandsValue = 8;
        private Boolean ClientDebugValue = false;
        private Boolean EnableLogNormalInValue = true;
        private Boolean EnableLogNormalOutValue = true;
        private Boolean EnableLogUnknownErrorValue = true;
        private Boolean EnableLogCriticalErrorValue = true;
        private Boolean EnableLogPerformanceValue = true;
        private Boolean EnableLogSystemValue = true;

        /// <summary>只能在启动前修改，以保证线程安全</summary>
        public int MaxBadCommands
        {
            get
            {
                return MaxBadCommandsValue;
            }
            set
            {
                if (IsRunning) { throw new InvalidOperationException(); }
                MaxBadCommandsValue = value;
            }
        }

        /// <summary>只能在启动前修改，以保证线程安全</summary>
        public Boolean ClientDebug
        {
            get
            {
                return ClientDebugValue;
            }
            set
            {
                if (IsRunning) { throw new InvalidOperationException(); }
                ClientDebugValue = value;
            }
        }

        /// <summary>只能在启动前修改，以保证线程安全</summary>
        public Boolean EnableLogNormalIn
        {
            get
            {
                return EnableLogNormalInValue;
            }
            set
            {
                if (IsRunning) { throw new InvalidOperationException(); }
                EnableLogNormalInValue = value;
            }
        }
        /// <summary>只能在启动前修改，以保证线程安全</summary>
        public Boolean EnableLogNormalOut
        {
            get
            {
                return EnableLogNormalOutValue;
            }
            set
            {
                if (IsRunning) { throw new InvalidOperationException(); }
                EnableLogNormalOutValue = value;
            }
        }
        /// <summary>只能在启动前修改，以保证线程安全</summary>
        public Boolean EnableLogUnknownError
        {
            get
            {
                return EnableLogUnknownErrorValue;
            }
            set
            {
                if (IsRunning) { throw new InvalidOperationException(); }
                EnableLogUnknownErrorValue = value;
            }
        }
        /// <summary>只能在启动前修改，以保证线程安全</summary>
        public Boolean EnableLogCriticalError
        {
            get
            {
                return EnableLogCriticalErrorValue;
            }
            set
            {
                if (IsRunning) { throw new InvalidOperationException(); }
                EnableLogCriticalErrorValue = value;
            }
        }
        /// <summary>只能在启动前修改，以保证线程安全</summary>
        public Boolean EnableLogPerformance
        {
            get
            {
                return EnableLogPerformanceValue;
            }
            set
            {
                if (IsRunning) { throw new InvalidOperationException(); }
                EnableLogPerformanceValue = value;
            }
        }
        /// <summary>只能在启动前修改，以保证线程安全</summary>
        public Boolean EnableLogSystem
        {
            get
            {
                return EnableLogSystemValue;
            }
            set
            {
                if (IsRunning) { throw new InvalidOperationException(); }
                EnableLogSystemValue = value;
            }
        }

        public ConcurrentDictionary<SessionContext, BinarySocketSession> SessionMappings = new ConcurrentDictionary<SessionContext, BinarySocketSession>();

        public BinarySocketServer()
        {
            ServerContext = new ServerContext();
            ServerContext.GetSessions = () => SessionMappings.Keys;
            si = new ServerImplementation(ServerContext);
            law = new JsonLogAspectWrapper<SessionContext>(si);
            law.ClientCommandIn += (c, CommandName, Parameters) =>
            {
                if (EnableLogNormalIn)
                {
                    var CommandLine = String.Format(@"/{0} {1}", CommandName, Parameters);
                    RaiseSessionLog(new SessionLogEntry { Token = c.SessionTokenString, RemoteEndPoint = c.RemoteEndPoint, Time = DateTime.UtcNow, Type = "In", Message = CommandLine });
                }
            };
            law.ClientCommandOut += (c, CommandName, Parameters) =>
            {
                if (EnableLogNormalOut)
                {
                    var CommandLine = String.Format(@"/svr {0} {1}", CommandName, Parameters);
                    RaiseSessionLog(new SessionLogEntry { Token = c.SessionTokenString, RemoteEndPoint = c.RemoteEndPoint, Time = DateTime.UtcNow, Type = "Out", Message = CommandLine });
                }
            };
            law.ServerCommand += (c, CommandName, Parameters) =>
            {
                if (EnableLogNormalOut)
                {
                    var CommandLine = String.Format(@"/svr {0} {1}", CommandName, Parameters);
                    RaiseSessionLog(new SessionLogEntry { Token = c.SessionTokenString, RemoteEndPoint = c.RemoteEndPoint, Time = DateTime.UtcNow, Type = "Out", Message = CommandLine });
                }
            };

            InnerServer = new BinaryServer<SessionContext>(law);
            InnerServer.ServerEvent += OnServerEvent;
            ServerContext.SchemaHash = InnerServer.Hash.ToString("X16", System.Globalization.CultureInfo.InvariantCulture);

            base.MaxConnectionsExceeded += OnMaxConnectionsExceeded;
            base.MaxConnectionsPerIPExceeded += OnMaxConnectionsPerIPExceeded;
        }

        public void RaiseError(SessionContext c, String CommandName, String Message)
        {
            si.RaiseError(c, CommandName, Message);
        }

        private void OnServerEvent(SessionContext c, String CommandName, UInt32 CommandHash, Byte[] Parameters)
        {
            BinarySocketSession Session = null;
            SessionMappings.TryGetValue(c, out Session);
            if (Session != null)
            {
                Session.WriteCommand(CommandName, CommandHash, Parameters);
            }
        }

        public event Action<SessionLogEntry> SessionLog;
        public void RaiseSessionLog(SessionLogEntry Entry)
        {
            if (SessionLog != null)
            {
                SessionLog(Entry);
            }
        }

        private void OnMaxConnectionsExceeded(BinarySocketSession s)
        {
            if (s != null && s.IsRunning)
            {
                s.RaiseError("", "Client host rejected: too many connections, please try again later.");
            }
        }
        private void OnMaxConnectionsPerIPExceeded(BinarySocketSession s)
        {
            if (s != null && s.IsRunning)
            {
                s.RaiseError("", "Client host rejected: too many connections from your IP(" + s.RemoteEndPoint.Address + "), please try again later.");
            }
        }
    }
}

File: trunk/Yuki/Examples/Communication/CPP/Src/Server/Servers/BinarySocketServer.cpp
#include "Servers/BinarySocketSession.h"
#include "Servers/BinarySocketServer.h"

#include "Utility.h"

namespace Server
{
    std::shared_ptr<Communication::Binary::BinaryServer<SessionContext>> BinarySocketServer::InnerServer() { return WorkPartInstance->Value().bs; }

    std::shared_ptr<Communication::Net::TcpSession> BinarySocketServer::CreateSession()
    {
        auto s = std::make_shared<BinarySocketSession>(IoService);
        s->Server = this->shared_from_this();
        return s;
    }

    int BinarySocketServer::GetMaxBadCommands() const
    {
        return MaxBadCommandsValue;
    }
    /// <summary>只能在启动前修改，以保证线程安全</summary>
    void BinarySocketServer::SetMaxBadCommands(int value)
    {
        if (IsRunning()) { throw std::logic_error("InvalidOperationException"); }
        MaxBadCommandsValue = value;
    }

    bool BinarySocketServer::GetClientDebug() const
    {
        return ClientDebugValue;
    }
    /// <summary>只能在启动前修改，以保证线程安全</summary>
    void BinarySocketServer::SetClientDebug(bool value)
    {
        if (IsRunning()) { throw std::logic_error("InvalidOperationException"); }
        ClientDebugValue = value;
    }

    bool BinarySocketServer::GetEnableLogNormalIn() const
    {
        return EnableLogNormalInValue;
    }
    /// <summary>只能在启动前修改，以保证线程安全</summary>
    void BinarySocketServer::SetEnableLogNormalIn(bool value)
    {
        if (IsRunning()) { throw std::logic_error("InvalidOperationException"); }
        EnableLogNormalInValue = value;
    }

    bool BinarySocketServer::GetEnableLogNormalOut() const
    {
        return EnableLogNormalOutValue;
    }
    /// <summary>只能在启动前修改，以保证线程安全</summary>
    void BinarySocketServer::SetEnableLogNormalOut(bool value)
    {
        if (IsRunning()) { throw std::logic_error("InvalidOperationException"); }
        EnableLogNormalOutValue = value;
    }

    bool BinarySocketServer::GetEnableLogUnknownError() const
    {
        return EnableLogUnknownErrorValue;
    }
    /// <summary>只能在启动前修改，以保证线程安全</summary>
    void BinarySocketServer::SetEnableLogUnknownError(bool value)
    {
        if (IsRunning()) { throw std::logic_error("InvalidOperationException"); }
        EnableLogUnknownErrorValue = value;
    }

    bool BinarySocketServer::GetEnableLogCriticalError() const
    {
        return EnableLogCriticalErrorValue;
    }
    /// <summary>只能在启动前修改，以保证线程安全</summary>
    void BinarySocketServer::SetEnableLogCriticalError(bool value)
    {
        if (IsRunning()) { throw std::logic_error("InvalidOperationException"); }
        EnableLogCriticalErrorValue = value;
    }

    bool BinarySocketServer::GetEnableLogPerformance() const
    {
        return EnableLogPerformanceValue;
    }
    /// <summary>只能在启动前修改，以保证线程安全</summary>
    void BinarySocketServer::SetEnableLogPerformance(bool value)
    {
        if (IsRunning()) { throw std::logic_error("InvalidOperationException"); }
        EnableLogPerformanceValue = value;
    }

    bool BinarySocketServer::GetEnableLogSystem() const
    {
        return EnableLogSystemValue;
    }
    /// <summary>只能在启动前修改，以保证线程安全</summary>
    void BinarySocketServer::SetEnableLogSystem(bool value)
    {
        if (IsRunning()) { throw std::logic_error("InvalidOperationException"); }
        EnableLogSystemValue = value;
    }

    BinarySocketServer::BinarySocketServer(boost::asio::io_service &IoService)
        : Communication::Net::TcpServer(IoService),
          WorkPartInstance(nullptr),
          MaxBadCommandsValue(8),
          ClientDebugValue(false),
          EnableLogNormalInValue(true),
          EnableLogNormalOutValue(true),
          EnableLogUnknownErrorValue(true),
          EnableLogCriticalErrorValue(true),
          EnableLogPerformanceValue(true),
          EnableLogSystemValue(true),
          SessionMappings(std::make_shared<TSessionMapping>())
    {
        sc = std::make_shared<ServerContext>();
        sc->GetSessions = [&]() -> std::shared_ptr<std::vector<std::shared_ptr<SessionContext>>>
        {
            return SessionMappings.Check<std::shared_ptr<std::vector<std::shared_ptr<SessionContext>>>>([&](const std::shared_ptr<TSessionMapping> &Mappings) -> std::shared_ptr<std::vector<std::shared_ptr<SessionContext>>>
            {
                auto l = std::make_shared<std::vector<std::shared_ptr<SessionContext>>>();
                for (auto i = Mappings->begin(); i != Mappings->end(); i.operator ++())
                {
                    l->push_back(std::get<0>(*i));
                }
                return l;
            });
        };

        auto OnServerEventHandler = [=](SessionContext &c, std::wstring CommandName, std::uint32_t CommandHash, std::shared_ptr<std::vector<std::uint8_t>> Parameters) { OnServerEvent(c, CommandName, CommandHash, Parameters); };

        WorkPartInstance = std::make_shared<Communication::BaseSystem::ThreadLocalVariable<WorkPart>>([=]() -> WorkPart *
        {
            auto si = std::make_shared<ServerImplementation>(sc);

            auto srv = std::make_shared<Communication::Binary::BinaryServer<SessionContext>>(si);
            srv->ServerEvent = OnServerEventHandler;

            auto i = new WorkPart();
            i->si = si;
            i->bs = srv;

            return i;
        });
        sc->SchemaHash = (boost::wformat(L"%16X") % InnerServer()->Hash()).str();

        MaxConnectionsExceeded = [=](std::shared_ptr<Communication::Net::TcpSession> s) { OnMaxConnectionsExceeded(std::static_pointer_cast<BinarySocketSession>(s)); };
        MaxConnectionsPerIPExceeded = [=](std::shared_ptr<Communication::Net::TcpSession> s) { OnMaxConnectionsExceeded(std::static_pointer_cast<BinarySocketSession>(s)); };
    }

    void BinarySocketServer::RaiseError(SessionContext &c, std::wstring CommandName, std::wstring Message)
    {
        WorkPartInstance->Value().si->RaiseError(c, CommandName, Message);
    }

    void BinarySocketServer::RaiseSessionLog(std::shared_ptr<SessionLogEntry> Entry)
    {
        if (SessionLog != nullptr)
        {
            SessionLog(Entry);
        }
    }

    void BinarySocketServer::OnServerEvent(SessionContext &c, std::wstring CommandName, std::uint32_t CommandHash, std::shared_ptr<std::vector<std::uint8_t>> Parameters)
    {
        std::shared_ptr<BinarySocketSession> Session = SessionMappings.Check<std::shared_ptr<BinarySocketSession>>([&](const std::shared_ptr<TSessionMapping> &Mappings) -> std::shared_ptr<BinarySocketSession>
        {
            auto cc = c.shared_from_this();
            if (Mappings->count(cc) > 0)
            {
                return (*Mappings)[cc];
            }
            else
            {
                return nullptr;
            }
        });

        if (Session != nullptr)
        {
            Session->WriteCommand(CommandName, CommandHash, Parameters);
        }
    }

    void BinarySocketServer::OnMaxConnectionsExceeded(std::shared_ptr<BinarySocketSession> s)
    {
        if (s != nullptr && s->IsRunning())
        {
            s->RaiseError(L"", L"Client host rejected: too many connections, please try again later.");
        }
    }

    void BinarySocketServer::OnMaxConnectionsPerIPExceeded(std::shared_ptr<BinarySocketSession> s)
    {
        if (s != nullptr && s->IsRunning())
        {
            s->RaiseError(L"", L"Client host rejected: too many connections from your IP(" + s2w(s->RemoteEndPoint.address().to_string()) + L"), please try again later.");
        }
    }
}
